{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# TCR-seq protocol"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "By Roman Sasik (rsasik@ucsd.edu) \n",
    "This Notebook describes the sequence of commands used in TCR-seq analysis. \n",
    "\n",
    "The multiplexing barcodes are assumed to follow the design described in this paper: \n",
    "_\"Linking T-cell receptor sequence to functional phenotype at the single-cell level\",_ A Han, J Glanville and MD Davis, Nature Biotechnology, 2014, 32 (7), p.684-92\n",
    "\n",
    "In addition to original perl scripts below, you need to install the superfast TCR repertoir processing java program `mitcr.jar`, which can be downloaded at http://mitcr.milaboratory.com/. The relevant paper is _MiTCR: software for T-cell receptor sequencing data analysis_ by DA Bolotin _et al._, Nature Methods 10, 813-814 (2013).\n",
    "\n",
    "Perl and java are assumed to be installed. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Demultiplexing TCR reads"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Processing starts with demultiplexing the reads from a single pair of large fastq files:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "!perl demultiplex_fastq_TCRplates.pl Sample_S1_L001_R1_001.fastq Sample_S1_L001_R2_001.fastq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "01A01A.fastq  01B08B.fastq  01D04A.fastq  01E11B.fastq\t01G07A.fastq\r\n",
      "01A01B.fastq  01B09A.fastq  01D04B.fastq  01E12A.fastq\t01G07B.fastq\r\n",
      "01A02A.fastq  01B09B.fastq  01D05A.fastq  01E12B.fastq\t01G08A.fastq\r\n",
      "01A02B.fastq  01B10A.fastq  01D05B.fastq  01F01A.fastq\t01G08B.fastq\r\n",
      "01A03A.fastq  01B10B.fastq  01D06A.fastq  01F01B.fastq\t01G09A.fastq\r\n",
      "01A03B.fastq  01B11A.fastq  01D06B.fastq  01F02A.fastq\t01G09B.fastq\r\n",
      "01A04A.fastq  01B11B.fastq  01D07A.fastq  01F02B.fastq\t01G10A.fastq\r\n",
      "01A04B.fastq  01B12A.fastq  01D07B.fastq  01F03A.fastq\t01G10B.fastq\r\n",
      "01A05A.fastq  01B12B.fastq  01D08A.fastq  01F03B.fastq\t01G11A.fastq\r\n",
      "01A05B.fastq  01C01A.fastq  01D08B.fastq  01F04A.fastq\t01G11B.fastq\r\n",
      "01A06A.fastq  01C01B.fastq  01D09A.fastq  01F04B.fastq\t01G12A.fastq\r\n",
      "01A06B.fastq  01C02A.fastq  01D09B.fastq  01F05A.fastq\t01G12B.fastq\r\n",
      "01A07A.fastq  01C02B.fastq  01D10A.fastq  01F05B.fastq\t01H01A.fastq\r\n",
      "01A07B.fastq  01C03A.fastq  01D10B.fastq  01F06A.fastq\t01H01B.fastq\r\n",
      "01A08A.fastq  01C03B.fastq  01D11A.fastq  01F06B.fastq\t01H02A.fastq\r\n",
      "01A08B.fastq  01C04A.fastq  01D11B.fastq  01F07A.fastq\t01H02B.fastq\r\n",
      "01A09A.fastq  01C04B.fastq  01D12A.fastq  01F07B.fastq\t01H03A.fastq\r\n",
      "01A09B.fastq  01C05A.fastq  01D12B.fastq  01F08A.fastq\t01H03B.fastq\r\n",
      "01A10A.fastq  01C05B.fastq  01E01A.fastq  01F08B.fastq\t01H04A.fastq\r\n",
      "01A10B.fastq  01C06A.fastq  01E01B.fastq  01F09A.fastq\t01H04B.fastq\r\n",
      "01A11A.fastq  01C06B.fastq  01E02A.fastq  01F09B.fastq\t01H05A.fastq\r\n",
      "01A11B.fastq  01C07A.fastq  01E02B.fastq  01F10A.fastq\t01H05B.fastq\r\n",
      "01A12A.fastq  01C07B.fastq  01E03A.fastq  01F10B.fastq\t01H06A.fastq\r\n",
      "01A12B.fastq  01C08A.fastq  01E03B.fastq  01F11A.fastq\t01H06B.fastq\r\n",
      "01B01A.fastq  01C08B.fastq  01E04A.fastq  01F11B.fastq\t01H07A.fastq\r\n",
      "01B01B.fastq  01C09A.fastq  01E04B.fastq  01F12A.fastq\t01H07B.fastq\r\n",
      "01B02A.fastq  01C09B.fastq  01E05A.fastq  01F12B.fastq\t01H08A.fastq\r\n",
      "01B02B.fastq  01C10A.fastq  01E05B.fastq  01G01A.fastq\t01H08B.fastq\r\n",
      "01B03A.fastq  01C10B.fastq  01E06A.fastq  01G01B.fastq\t01H09A.fastq\r\n",
      "01B03B.fastq  01C11A.fastq  01E06B.fastq  01G02A.fastq\t01H09B.fastq\r\n",
      "01B04A.fastq  01C11B.fastq  01E07A.fastq  01G02B.fastq\t01H10A.fastq\r\n",
      "01B04B.fastq  01C12A.fastq  01E07B.fastq  01G03A.fastq\t01H10B.fastq\r\n",
      "01B05A.fastq  01C12B.fastq  01E08A.fastq  01G03B.fastq\t01H11A.fastq\r\n",
      "01B05B.fastq  01D01A.fastq  01E08B.fastq  01G04A.fastq\t01H11B.fastq\r\n",
      "01B06A.fastq  01D01B.fastq  01E09A.fastq  01G04B.fastq\t01H12A.fastq\r\n",
      "01B06B.fastq  01D02A.fastq  01E09B.fastq  01G05A.fastq\t01H12B.fastq\r\n",
      "01B07A.fastq  01D02B.fastq  01E10A.fastq  01G05B.fastq\r\n",
      "01B07B.fastq  01D03A.fastq  01E10B.fastq  01G06A.fastq\r\n",
      "01B08A.fastq  01D03B.fastq  01E11A.fastq  01G06B.fastq\r\n"
     ]
    }
   ],
   "source": [
    "!ls *[A,B].fastq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This script demultiplexes reads multiplexed in a single pair of large fastq files and saves them into separate fastq files whose names indicate Plate, Well, and TCR isoform (A or B), for instance 01H12B.fastq. \n",
    "Up to _one mismatch_ is allowed in any of the Plate, Well Row, Well Column, and TCR Isoform barcodes.\n",
    "\n",
    "It will create 2x96 files (one per TCR isoform) per each Plate (a lot of files!)\n",
    "\n",
    "This script will ignore all reads from plates whose code is commented out (see below in source code). This is useful when there is a mixture of TCR genotyping reads and phenotyping reads. \n",
    "\n",
    "There is a separate demultiplex script for the phenotyping reads (see below). This is `demultiplex_fastq_TCRplates.pl`:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#!/usr/bin/perl\n",
    "$fileR1 = $ARGV[0];\n",
    "$fileR2 = $ARGV[1];\n",
    "\n",
    "open(F1,$fileR1);\n",
    "open(F2,$fileR2);\n",
    "\n",
    "%plate = (\n",
    "   \"GCAGA\" => \"01\", #uncomment this line if plate code 01 is among the sequences to be demultiplexed\n",
    "#   \"TCGAA\" => \"02\",\n",
    "#   \"AACAA\" => \"03\",\n",
    "#   \"GGTGC\" => \"04\",\n",
    "#   \"TTGGT\" => \"05\",\n",
    "#   \"CATTC\" => \"06\",\n",
    "#   \"ATTGG\" => \"07\",\n",
    "#   \"CGGTT\" => \"08\",\n",
    "#   \"ATCCT\" => \"09\",\n",
    "#   \"ATGTC\" => \"10\",\n",
    "#   \"TCACG\" => \"11\",\n",
    "#   \"AGACC\" => \"12\",\n",
    "#   \"CCCCA\" => \"13\",\n",
    "#   \"GCGCT\" => \"14\",\n",
    "#   \"TCCTT\" => \"15\",\n",
    "#   \"TATAT\" => \"16\",\n",
    "#   \"CGTAA\" => \"17\",\n",
    "#   \"AAGGT\" => \"18\",\n",
    "#   \"AGCTC\" => \"19\",\n",
    "#   \"CTTGC\" => \"20\",\n",
    "#   \"GTATC\" => \"21\",\n",
    "#   \"TATGA\" => \"22\",\n",
    "#   \"CACAC\" => \"23\",\n",
    "#   \"ACACT\" => \"24\",\n",
    "#   \"ACTAC\" => \"25\",\n",
    "#   \"GTTAC\" => \"26\",\n",
    ");\n",
    "%row = ( #if you want output for all rows, leave them all uncommented\n",
    "   \"TAAGC\" => \"A\", \n",
    "   \"TGCAC\" => \"B\",\n",
    "   \"CTCAG\" => \"C\",\n",
    "   \"GGAAT\" => \"D\",\n",
    "   \"CGAGG\" => \"E\",\n",
    "   \"AGGAG\" => \"F\",\n",
    "   \"TGTTG\" => \"G\",\n",
    "   \"CAACT\" => \"H\",\n",
    ");\n",
    "%col = ( #if you want output for all columns, leave them all uncommented\n",
    "   \"GTTCA\" => \"01\", \n",
    "   \"CAGGA\" => \"02\",\n",
    "   \"TTATA\" => \"03\",\n",
    "   \"CCTGT\" => \"04\",\n",
    "   \"ACCGC\" => \"05\",\n",
    "   \"ACTTA\" => \"06\",\n",
    "   \"GCTAG\" => \"07\",\n",
    "   \"GACGT\" => \"08\",\n",
    "   \"GGCTA\" => \"09\",\n",
    "   \"GAATG\" => \"10\",\n",
    "   \"CCAAC\" => \"11\",\n",
    "   \"GAGAC\" => \"12\",\n",
    ");\n",
    "%TCR = (\n",
    "   \"GTCAC\" => \"A\", # TCRA\n",
    "   \"GAGAT\" => \"B\",\n",
    ");\n",
    "\n",
    "foreach $plateID (keys(%plate)) {\n",
    "   foreach $rowID (keys(%row)) {\n",
    "      foreach $colID (keys(%col)) {\n",
    "         foreach $TCRID (keys(%TCR)) {\n",
    "            $fh = $plate{$plateID}.$row{$rowID}.$col{$colID}.$TCR{$TCRID};\n",
    "            open $fh, '>', $fh.\".fastq\"; #open file for writing at the end\n",
    "         }\n",
    "      }\n",
    "   }\n",
    "}\n",
    "\n",
    "while($A1 = <F1>) { #read 4 lines from R1 and 4 lines from R2\n",
    "   $A2 = <F1>;\n",
    "   $A3 = <F1>;\n",
    "   $A4 = <F1>;\n",
    "\n",
    "   $B1 = <F2>;\n",
    "   $B2 = <F2>;\n",
    "   $B3 = <F2>;\n",
    "   $B4 = <F2>;\n",
    "\n",
    "   $ID = substr($A2, 2, 5); #plate ID barcode\n",
    "#  now find what the true bar code should have been if imperfect match\n",
    "   $score = 0;\n",
    "   $trueID = \"\";\n",
    "   foreach $key (keys(%plate)) {\n",
    "      my $count = ($ID^$key)  =~ tr/\\0//;\n",
    "      if ($count > $score) {\n",
    "         $score = $count;\n",
    "         $trueID = $key\n",
    "      }\n",
    "   }\n",
    "   if ($score >= 4) {#accept $true_plateID as the true plate ID\n",
    "      $rowID = $trueID;\n",
    "   } else {#leave $plateID blank - sequence won't be output\n",
    "      $rowID = \"\"\n",
    "   }\n",
    "\n",
    "   $ID = substr($B2, 2, 5); #column ID\n",
    "#  now find what the true bar code should have been if imperfect match\n",
    "   $score = 0;\n",
    "   $trueID = \"\";\n",
    "   foreach $key (keys(%col)) {\n",
    "      my $count = ($ID^$key)  =~ tr/\\0//;\n",
    "      if ($count > $score) {\n",
    "         $score = $count;\n",
    "         $trueID = $key\n",
    "      }\n",
    "   }\n",
    "   if ($score >= 4) {#accept $true_plateID as the true plate ID\n",
    "      $colID = $trueID;\n",
    "   } else {#leave $plateID blank - sequence won't be output\n",
    "      $colID = \"\"\n",
    "   }\n",
    "\n",
    "   $ID = substr($B2, 7, 5); #TCR ID\n",
    "#  now find what the true bar code should have been if imperfect match\n",
    "   $score = 0;\n",
    "   $trueID = \"\";\n",
    "   foreach $key (keys(%TCR)) {\n",
    "      my $count = ($ID^$key)  =~ tr/\\0//;\n",
    "      if ($count > $score) {\n",
    "         $score = $count;\n",
    "         $trueID = $key\n",
    "      }\n",
    "   }\n",
    "   if ($score >= 4) {\n",
    "      $TCRID = $trueID;\n",
    "   } else {\n",
    "      $TCRID = \"\"\n",
    "   }\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "   if (exists $plate{$plateID}  and exists $row{$rowID}  and exists $col{$colID} and exists $TCR{$TCRID}) {\n",
    "      $fh = $plate{$plateID}.$row{$rowID}.$col{$colID}.$TCR{$TCRID};\n",
    "      print $fh $A1.$A2.$A3.$A4.$B1.$B2.$B3.$B4;\n",
    "   };\n",
    "}\n",
    "close(F1);\n",
    "close(F2);\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Analyzing demultiplexed fastq files for TCRA/B species"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "After demultiplexing, each individual fastq file will be processed by `mitcr`. The output is a separate result file for each well, e.g., `01A06A_result.txt`. The example below will produce reports for plate 01, row A and columns 06 through 09 (see source code below). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "01A06B\n",
      "Initialisation: progress unknown\n",
      "01A06A\n",
      "Initialisation: progress unknown\n",
      "01A08B\n",
      "Initialisation: progress unknown\n",
      "01A08A\n",
      "Initialisation: progress unknown\n",
      "01A07B\n",
      "Initialisation: progress unknown\n",
      "01A07A\n",
      "Initialisation: progress unknown\n",
      "01A09B\n",
      "Initialisation: progress unknown\n",
      "01A09A\n",
      "Initialisation: progress unknown\n",
      "01A06A_result.txt  01A07A_result.txt  01A08A_result.txt  01A09A_result.txt\n",
      "01A06B_result.txt  01A07B_result.txt  01A08B_result.txt  01A09B_result.txt\n"
     ]
    }
   ],
   "source": [
    "!perl analyze_wells.pl\n",
    "!ls *_result.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The output is a tab-delimited file whose main components are these (this is the content of file 01A06A_result.txt):\n",
    "\n",
    "<img src = \"files/TCRA.png\">\n",
    "\n",
    "The first column is the number of times this sequence is seen; the second column is the fraction (not a percentage) of the total count of sequences in the well. This is especially useful when there are two species of TCRA expressed in a single cell (as in this case). It does not happen with TCRB.The v- j- and d- alleles of the TCR are listed. The last two lines (a tiny fraction of the number of reads) are a result of sequencing/PCR errors. The program _mitcr_ has an error-checking algorithm that reduces these calls. For details see _MiTCR: software for T-cell receptor sequencing data analysis_ by DA Bolotin _et al._, Nature Methods 10, 813-814 (2013).\n",
    "\n",
    "This is the source of `analyze_wells.pl`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#!/usr/bin/perl\n",
    "\n",
    "%plate = (\n",
    "   \"GCAGA\" => \"01\",\n",
    "#   \"TCGAA\" => \"02\",\n",
    "#   \"AACAA\" => \"03\",\n",
    "#   \"GGTGC\" => \"04\",\n",
    "#   \"TTGGT\" => \"05\",\n",
    "#   \"CATTC\" => \"06\",\n",
    "#   \"ATTGG\" => \"07\",\n",
    "#   \"CGGTT\" => \"08\",\n",
    "#   \"ATCCT\" => \"09\",\n",
    "#   \"ATGTC\" => \"10\",\n",
    "#   \"TCACG\" => \"11\",\n",
    "#   \"AGACC\" => \"12\",\n",
    "#   \"CCCCA\" => \"13\",\n",
    "#   \"GCGCT\" => \"14\",\n",
    "#   \"TCCTT\" => \"15\",\n",
    "#   \"TATAT\" => \"16\",\n",
    "#   \"CGTAA\" => \"17\",\n",
    "#   \"AAGGT\" => \"18\",\n",
    "#   \"AGCTC\" => \"19\",\n",
    "#   \"CTTGC\" => \"20\",\n",
    "#   \"GTATC\" => \"21\",\n",
    "#   \"TATGA\" => \"22\",\n",
    "#   \"CACAC\" => \"23\",\n",
    "#   \"ACACT\" => \"24\",\n",
    "#   \"ACTAC\" => \"25\",\n",
    "#   \"GTTAC\" => \"26\",\n",
    ");\n",
    "%row = ( #uncomment line if you want output for row A, etc.\n",
    "   \"TAAGC\" => \"A\", \n",
    "#   \"TGCAC\" => \"B\",\n",
    "#   \"CTCAG\" => \"C\",\n",
    "#   \"GGAAT\" => \"D\",\n",
    "#   \"CGAGG\" => \"E\",\n",
    "#   \"AGGAG\" => \"F\",\n",
    "#   \"TGTTG\" => \"G\",\n",
    "#   \"CAACT\" => \"H\",\n",
    ");\n",
    "%col = ( #uncomment line if you want output for column 01, etc.\n",
    "#   \"GTTCA\" => \"01\",\n",
    "#   \"CAGGA\" => \"02\",\n",
    "#   \"TTATA\" => \"03\",\n",
    "#   \"CCTGT\" => \"04\",\n",
    "#   \"ACCGC\" => \"05\",\n",
    "   \"ACTTA\" => \"06\",\n",
    "   \"GCTAG\" => \"07\",\n",
    "   \"GACGT\" => \"08\",\n",
    "   \"GGCTA\" => \"09\",\n",
    "#   \"GAATG\" => \"10\",\n",
    "#   \"CCAAC\" => \"11\",\n",
    "#   \"GAGAC\" => \"12\",\n",
    ");\n",
    "%TCR = (\n",
    "   \"GTCAC\" => \"A\", # TCRA\n",
    "   \"GAGAT\" => \"B\",\n",
    ");\n",
    "\n",
    "\n",
    "foreach $plateID (sort (keys(%plate))) {\n",
    "   foreach $rowID (sort (keys(%row))) {\n",
    "      foreach $colID (sort (keys(%col))) {\n",
    "         foreach $TCRID (sort (keys(%TCR))) {\n",
    "            $fh = $plate{$plateID}.$row{$rowID}.$col{$colID}.$TCR{$TCRID};\n",
    "            print \"$fh\\n\";\n",
    "            system(\"java -Xmx10g -jar ./mitcr.jar -pset flex -gene TR$TCR{$TCRID} $fh.fastq $fh\\_result.txt\")\n",
    "         }\n",
    "      }\n",
    "   }\n",
    "}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Demultiplexing phenotyping reads"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The following command demultiplexes _phenotyping_ reads multiplexed in a single pair of large fastq files and saves them into separate fastq files whose names indicate Plate, Well, and \"R1\" or \"R2\" for left or right read, for instance 03H12R1.fastq. Up to one mismatch is allowed in any of the Plate, Well Row, or Well Column barcodes.\n",
    "\n",
    "It will create 2x96 files per each Plate.\n",
    "\n",
    "This script will ignore all reads from plates whose code is commented out (see below in source code). This is useful when there is a mixture of TCR genotyping reads and phenotyping reads."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "03A01R1.fastq  03B08R2.fastq  03D04R1.fastq  03E11R2.fastq  03G07R1.fastq\r\n",
      "03A01R2.fastq  03B09R1.fastq  03D04R2.fastq  03E12R1.fastq  03G07R2.fastq\r\n",
      "03A02R1.fastq  03B09R2.fastq  03D05R1.fastq  03E12R2.fastq  03G08R1.fastq\r\n",
      "03A02R2.fastq  03B10R1.fastq  03D05R2.fastq  03F01R1.fastq  03G08R2.fastq\r\n",
      "03A03R1.fastq  03B10R2.fastq  03D06R1.fastq  03F01R2.fastq  03G09R1.fastq\r\n",
      "03A03R2.fastq  03B11R1.fastq  03D06R2.fastq  03F02R1.fastq  03G09R2.fastq\r\n",
      "03A04R1.fastq  03B11R2.fastq  03D07R1.fastq  03F02R2.fastq  03G10R1.fastq\r\n",
      "03A04R2.fastq  03B12R1.fastq  03D07R2.fastq  03F03R1.fastq  03G10R2.fastq\r\n",
      "03A05R1.fastq  03B12R2.fastq  03D08R1.fastq  03F03R2.fastq  03G11R1.fastq\r\n",
      "03A05R2.fastq  03C01R1.fastq  03D08R2.fastq  03F04R1.fastq  03G11R2.fastq\r\n",
      "03A06R1.fastq  03C01R2.fastq  03D09R1.fastq  03F04R2.fastq  03G12R1.fastq\r\n",
      "03A06R2.fastq  03C02R1.fastq  03D09R2.fastq  03F05R1.fastq  03G12R2.fastq\r\n",
      "03A07R1.fastq  03C02R2.fastq  03D10R1.fastq  03F05R2.fastq  03H01R1.fastq\r\n",
      "03A07R2.fastq  03C03R1.fastq  03D10R2.fastq  03F06R1.fastq  03H01R2.fastq\r\n",
      "03A08R1.fastq  03C03R2.fastq  03D11R1.fastq  03F06R2.fastq  03H02R1.fastq\r\n",
      "03A08R2.fastq  03C04R1.fastq  03D11R2.fastq  03F07R1.fastq  03H02R2.fastq\r\n",
      "03A09R1.fastq  03C04R2.fastq  03D12R1.fastq  03F07R2.fastq  03H03R1.fastq\r\n",
      "03A09R2.fastq  03C05R1.fastq  03D12R2.fastq  03F08R1.fastq  03H03R2.fastq\r\n",
      "03A10R1.fastq  03C05R2.fastq  03E01R1.fastq  03F08R2.fastq  03H04R1.fastq\r\n",
      "03A10R2.fastq  03C06R1.fastq  03E01R2.fastq  03F09R1.fastq  03H04R2.fastq\r\n",
      "03A11R1.fastq  03C06R2.fastq  03E02R1.fastq  03F09R2.fastq  03H05R1.fastq\r\n",
      "03A11R2.fastq  03C07R1.fastq  03E02R2.fastq  03F10R1.fastq  03H05R2.fastq\r\n",
      "03A12R1.fastq  03C07R2.fastq  03E03R1.fastq  03F10R2.fastq  03H06R1.fastq\r\n",
      "03A12R2.fastq  03C08R1.fastq  03E03R2.fastq  03F11R1.fastq  03H06R2.fastq\r\n",
      "03B01R1.fastq  03C08R2.fastq  03E04R1.fastq  03F11R2.fastq  03H07R1.fastq\r\n",
      "03B01R2.fastq  03C09R1.fastq  03E04R2.fastq  03F12R1.fastq  03H07R2.fastq\r\n",
      "03B02R1.fastq  03C09R2.fastq  03E05R1.fastq  03F12R2.fastq  03H08R1.fastq\r\n",
      "03B02R2.fastq  03C10R1.fastq  03E05R2.fastq  03G01R1.fastq  03H08R2.fastq\r\n",
      "03B03R1.fastq  03C10R2.fastq  03E06R1.fastq  03G01R2.fastq  03H09R1.fastq\r\n",
      "03B03R2.fastq  03C11R1.fastq  03E06R2.fastq  03G02R1.fastq  03H09R2.fastq\r\n",
      "03B04R1.fastq  03C11R2.fastq  03E07R1.fastq  03G02R2.fastq  03H10R1.fastq\r\n",
      "03B04R2.fastq  03C12R1.fastq  03E07R2.fastq  03G03R1.fastq  03H10R2.fastq\r\n",
      "03B05R1.fastq  03C12R2.fastq  03E08R1.fastq  03G03R2.fastq  03H11R1.fastq\r\n",
      "03B05R2.fastq  03D01R1.fastq  03E08R2.fastq  03G04R1.fastq  03H11R2.fastq\r\n",
      "03B06R1.fastq  03D01R2.fastq  03E09R1.fastq  03G04R2.fastq  03H12R1.fastq\r\n",
      "03B06R2.fastq  03D02R1.fastq  03E09R2.fastq  03G05R1.fastq  03H12R2.fastq\r\n",
      "03B07R1.fastq  03D02R2.fastq  03E10R1.fastq  03G05R2.fastq\r\n",
      "03B07R2.fastq  03D03R1.fastq  03E10R2.fastq  03G06R1.fastq\r\n",
      "03B08R1.fastq  03D03R2.fastq  03E11R1.fastq  03G06R2.fastq\r\n"
     ]
    }
   ],
   "source": [
    "!perl demultiplex_fastq_phenoplates.pl Sample_S1_L001_R1_001.fastq Sample_S1_L001_R2_001.fastq\n",
    "!ls 03*.fastq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The source code of demultiplex_fastq_phenoplates.pl is here (in this example, Plate 03 contains phenotyping reads):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#!/usr/bin/perl\n",
    "\n",
    "$fileR1 = $ARGV[0];\n",
    "$fileR2 = $ARGV[1];\n",
    "\n",
    "open(F1,$fileR1);\n",
    "open(F2,$fileR2);\n",
    "\n",
    "%plate = (\n",
    "#   \"GCAGA\" => \"01\",\n",
    "#   \"TCGAA\" => \"02\",\n",
    "   \"AACAA\" => \"03\",\n",
    "#   \"GGTGC\" => \"04\",\n",
    "#   \"TTGGT\" => \"05\",\n",
    "#   \"CATTC\" => \"06\",\n",
    ");\n",
    "%row = (\n",
    "   \"TAAGC\" => \"A\",\n",
    "   \"TGCAC\" => \"B\",\n",
    "   \"CTCAG\" => \"C\",\n",
    "   \"GGAAT\" => \"D\",\n",
    "   \"CGAGG\" => \"E\",\n",
    "   \"AGGAG\" => \"F\",\n",
    "   \"TGTTG\" => \"G\",\n",
    "   \"CAACT\" => \"H\",\n",
    ");\n",
    "%col = (\n",
    "   \"GTTCA\" => \"01\",\n",
    "   \"CAGGA\" => \"02\",\n",
    "   \"TTATA\" => \"03\",\n",
    "   \"CCTGT\" => \"04\",\n",
    "   \"ACCGC\" => \"05\",\n",
    "   \"ACTTA\" => \"06\",\n",
    "   \"GCTAG\" => \"07\",\n",
    "   \"GACGT\" => \"08\",\n",
    "   \"GGCTA\" => \"09\",\n",
    "   \"GAATG\" => \"10\",\n",
    "   \"CCAAC\" => \"11\",\n",
    "   \"GAGAC\" => \"12\",\n",
    ");\n",
    "\n",
    "foreach $plateID (keys(%plate)) {\n",
    "   foreach $rowID (keys(%row)) {\n",
    "      foreach $colID (keys(%col)) {\n",
    "            $fh = $plate{$plateID}.$row{$rowID}.$col{$colID};\n",
    "            $fh1 = $plate{$plateID}.$row{$rowID}.$col{$colID}.\"1\";\n",
    "            $fh2 = $plate{$plateID}.$row{$rowID}.$col{$colID}.\"2\";\n",
    "            open $fh1, '>', $fh.\"R1.fastq\";\n",
    "            open $fh2, '>', $fh.\"R2.fastq\";\n",
    "      }\n",
    "   }\n",
    "}\n",
    "\n",
    "while($A1 = <F1>) { #read 4 lines from R1 and 4 lines from R2\n",
    "   $A2 = <F1>;\n",
    "   $A3 = <F1>;\n",
    "   $A4 = <F1>;\n",
    "\n",
    "   $B1 = <F2>;\n",
    "   $B2 = <F2>;\n",
    "   $B3 = <F2>;\n",
    "   $B4 = <F2>;\n",
    "\n",
    "#  now find out if the bar codes make sense\n",
    "\n",
    "   $ID = substr($A2, 2, 5); #plate ID\n",
    "#  now find what the true bar code should have been if imperfect match\n",
    "   $score = 0;\n",
    "   $trueID = \"\";\n",
    "   foreach $key (keys(%plate)) {\n",
    "      my $count = ($ID^$key)  =~ tr/\\0//;\n",
    "      if ($count > $score) {\n",
    "         $score = $count;\n",
    "         $trueID = $key\n",
    "      }\n",
    "   }\n",
    "   if ($score >= 4) {#accept $true_plateID as the true plate ID\n",
    "      $plateID = $trueID;\n",
    "   } else {#leave $plateID blank - sequence won't be output\n",
    "      $plateID = \"\"\n",
    "   }\n",
    "\n",
    "   $ID = substr($A2, 9, 5); #row ID\n",
    "#  now find what the true bar code should have been if imperfect match\n",
    "   $score = 0;\n",
    "   $trueID = \"\";\n",
    "   foreach $key (keys(%row)) {\n",
    "      my $count = ($ID^$key)  =~ tr/\\0//;\n",
    "      if ($count > $score) {\n",
    "         $score = $count;\n",
    "         $trueID = $key\n",
    "      }\n",
    "   }\n",
    "   if ($score >= 4) {\n",
    "      $rowID = $trueID;\n",
    "   } else {\n",
    "      $rowID = \"\"\n",
    "   }\n",
    "\n",
    "   $ID = substr($B2, 2, 5); #column ID\n",
    "#  now find what the true bar code should have been if imperfect match\n",
    "   $score = 0;\n",
    "   $trueID = \"\";\n",
    "   foreach $key (keys(%col)) {\n",
    "      my $count = ($ID^$key)  =~ tr/\\0//;\n",
    "      if ($count > $score) {\n",
    "         $score = $count;\n",
    "         $trueID = $key\n",
    "      }\n",
    "   }\n",
    "   if ($score >= 4) {\n",
    "      $colID = $trueID;\n",
    "   } else {\n",
    "      $colID = \"\"\n",
    "   }\n",
    "\n",
    "   if (exists $plate{$plateID}  and exists $row{$rowID}  and exists $col{$colID} ) {\n",
    "      $fh1 = $plate{$plateID}.$row{$rowID}.$col{$colID}.\"1\";\n",
    "      $fh2 = $plate{$plateID}.$row{$rowID}.$col{$colID}.\"2\";\n",
    "      print $fh1 $A1.$A2.$A3.$A4;\n",
    "      print $fh2 $B1.$B2.$B3.$B4;\n",
    "   };\n",
    "}\n",
    "close(F1);\n",
    "close(F2);\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Analyze demultiplexed phenotyping fastq files for expression levels of 17 cytokines and transcription factors"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The following command will produce expression counts for all 17 cytokines and TF's, separately for each well:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "03A01R1.count  03B09R1.count  03D05R1.count  03F01R1.count  03G09R1.count\r\n",
      "03A02R1.count  03B10R1.count  03D06R1.count  03F02R1.count  03G10R1.count\r\n",
      "03A03R1.count  03B11R1.count  03D07R1.count  03F03R1.count  03G11R1.count\r\n",
      "03A04R1.count  03B12R1.count  03D08R1.count  03F04R1.count  03G12R1.count\r\n",
      "03A05R1.count  03C01R1.count  03D09R1.count  03F05R1.count  03H01R1.count\r\n",
      "03A06R1.count  03C02R1.count  03D10R1.count  03F06R1.count  03H02R1.count\r\n",
      "03A07R1.count  03C03R1.count  03D11R1.count  03F07R1.count  03H03R1.count\r\n",
      "03A08R1.count  03C04R1.count  03D12R1.count  03F08R1.count  03H04R1.count\r\n",
      "03A09R1.count  03C05R1.count  03E01R1.count  03F09R1.count  03H05R1.count\r\n",
      "03A10R1.count  03C06R1.count  03E02R1.count  03F10R1.count  03H06R1.count\r\n",
      "03A11R1.count  03C07R1.count  03E03R1.count  03F11R1.count  03H07R1.count\r\n",
      "03A12R1.count  03C08R1.count  03E04R1.count  03F12R1.count  03H08R1.count\r\n",
      "03B01R1.count  03C09R1.count  03E05R1.count  03G01R1.count  03H09R1.count\r\n",
      "03B02R1.count  03C10R1.count  03E06R1.count  03G02R1.count  03H10R1.count\r\n",
      "03B03R1.count  03C11R1.count  03E07R1.count  03G03R1.count  03H11R1.count\r\n",
      "03B04R1.count  03C12R1.count  03E08R1.count  03G04R1.count  03H12R1.count\r\n",
      "03B05R1.count  03D01R1.count  03E09R1.count  03G05R1.count\r\n",
      "03B06R1.count  03D02R1.count  03E10R1.count  03G06R1.count\r\n",
      "03B07R1.count  03D03R1.count  03E11R1.count  03G07R1.count\r\n",
      "03B08R1.count  03D04R1.count  03E12R1.count  03G08R1.count\r\n"
     ]
    }
   ],
   "source": [
    "!perl count_cytokines.pl\n",
    "!ls *.count"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The output is a set of tab-delimited files such as 03F03R1.count. Only the R1 read is used for counting; the R2 read is redundant (and lower quality anyway). The content of this file looks something close to this:\n",
    "\n",
    "<img src = \"files/counts.png\">\n",
    "\n",
    "The source code of count_cytokines.pl is here (Plate 03 has pheno reads):"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#!/usr/bin/perl\n",
    "\n",
    "%plate = (\n",
    "#  \"GCAGA\" => \"01\",\n",
    "#  \"TCGAA\" => \"02\",\n",
    "   \"AACAA\" => \"03\",\n",
    "#  \"GGTGC\" => \"04\",\n",
    "#  \"TTGGT\" => \"05\",\n",
    "#  \"CATTC\" => \"06\",\n",
    ");\n",
    "%row = (\n",
    "   \"TAAGC\" => \"A\",\n",
    "   \"TGCAC\" => \"B\",\n",
    "   \"CTCAG\" => \"C\",\n",
    "   \"GGAAT\" => \"D\",\n",
    "   \"CGAGG\" => \"E\",\n",
    "   \"AGGAG\" => \"F\",\n",
    "   \"TGTTG\" => \"G\",\n",
    "   \"CAACT\" => \"H\",\n",
    ");\n",
    "%col = (\n",
    "   \"GTTCA\" => \"01\",\n",
    "   \"CAGGA\" => \"02\",\n",
    "   \"TTATA\" => \"03\",\n",
    "   \"CCTGT\" => \"04\",\n",
    "   \"ACCGC\" => \"05\",\n",
    "   \"ACTTA\" => \"06\",\n",
    "   \"GCTAG\" => \"07\",\n",
    "   \"GACGT\" => \"08\",\n",
    "   \"GGCTA\" => \"09\",\n",
    "   \"GAATG\" => \"10\",\n",
    "   \"CCAAC\" => \"11\",\n",
    "   \"GAGAC\" => \"12\",\n",
    ");\n",
    "%cyt = (\n",
    "   \"GCCGGAGGAGGTGGATGTGC\" => \"GATA3\",\n",
    "   \"CCCAACACAGGAGCGCACTG\" => \"TBET\",\n",
    "   \"GGCAGCCAAGGCCCTGTCGT\" => \"FOXP3\",\n",
    "   \"AGAGGAAGTCCATGTGGGAG\" => \"RORC\",\n",
    "   \"GCGAGCTGGTGCGCACCGAC\" => \"RUNX1\",\n",
    "   \"GGACCACGCAGGCGAGCTCG\" => \"RUNX3\",\n",
    "   \"CCTACACGGCCCCACCTGCC\" => \"BCL6\",\n",
    "   \"CCACAGAACTGAAACATCTT\" => \"IL2\",\n",
    "   \"CCCAAGCTGAGAACCAAGAC\" => \"IL10\",\n",
    "   \"AGACCTCTTTTATGATGGCC\" => \"IL12A\",\n",
    "   \"GGTATGGAGCATCAACCTGA\" => \"IL13\",\n",
    "   \"CAACCTGAACATCCATAACC\" => \"IL17A\",\n",
    "   \"GGGTTCTCTTGGCTGTTACT\" => \"IFNG\",\n",
    "   \"GGAGGCGCTCCCCAAGAAGA\" => \"TNFA\",\n",
    "   \"CCGAGAAGCGGTACCTGAAC\" => \"TGFB\",\n",
    "   \"GCCAACTTTGCAGCCCAGAA\" => \"PRF1\",\n",
    "   \"CCACAATATCAAAGAACAGG\" => \"GZMB\",\n",
    ");\n",
    "foreach $plateID (sort (keys(%plate))) {\n",
    "   foreach $rowID (sort (keys(%row))) {\n",
    "      foreach $colID (sort (keys(%col))) {\n",
    "         $fh = $plate{$plateID}.$row{$rowID}.$col{$colID};\n",
    "         open(F1,$fh.\"R1.fastq\");\n",
    "         open $fh, '>', $fh.\"R1.count\";\n",
    "         print $fh \"\\t$fh\\n\"; #print header\n",
    "\n",
    "#        zero out counters\n",
    "         foreach $key (keys(%cyt)) {$count{$cyt{$key}} = 0};\n",
    "         while($A1 = <F1>) { #read 4 lines from R1 and 4 lines from R2\n",
    "            $A2 = <F1>;\n",
    "            $A3 = <F1>;\n",
    "            $A4 = <F1>;\n",
    "#           now find out if the bar codes make sense\n",
    "            $seq = substr($A2, 36, 20);\n",
    "            if (exists $cyt{$seq}) {$count{$cyt{$seq}}++}; #add to count\n",
    "         };\n",
    "         foreach $key (keys(%cyt)) {\n",
    "            print $fh $cyt{$key}.\"\\t\".$count{$cyt{$key}}.\"\\n\"\n",
    "         };\n",
    "         close(F1);\n",
    "         close($fh);\n",
    "\n",
    "      }\n",
    "   }\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Cleanup after exercize:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "!rm 0*"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
